File: js/src/keypair.ts
/**
 * Key pair generation and management for TOPAY-Z512
 */

import { 
  PRIVATE_KEY_SIZE, 
  PUBLIC_KEY_SIZE, 
  PrivateKey, 
  PublicKey, 
  KeyPair,
  InvalidKeySizeError 
} from './index';
import { secureRandom, validateSize, copyBytes, constantTimeEqual, secureZero } from './utils';
import { computeHash, deriveKey } from './hash';

/**
 * Generates a new cryptographic key pair
 * @returns Promise resolving to a new key pair
 */
export async function generateKeyPair(): Promise<KeyPair> {
  // Generate private key from secure random bytes
  const privateKey = await secureRandom(PRIVATE_KEY_SIZE);
  
  // Derive public key from private key using hash-based derivation
  const publicKey = await derivePublicKey(privateKey);
  
  return {
    privateKey,
    publicKey
  };
}

/**
 * Derives a public key from a private key
 * @param privateKey - Private key to derive from
 * @returns Promise resolving to the derived public key
 * @throws InvalidKeySizeError if private key size is invalid
 */
export async function derivePublicKey(privateKey: PrivateKey): Promise<PublicKey> {
  validateSize(privateKey, PRIVATE_KEY_SIZE, 'private key');
  
  // Use hash-based public key derivation for quantum safety
  const publicKey = await computeHash(privateKey);
  validateSize(publicKey, PUBLIC_KEY_SIZE, 'public key');
  
  return publicKey;
}

/**
 * Generates a key pair from a seed
 * @param seed - Seed bytes for deterministic generation
 * @returns Promise resolving to a key pair
 */
export async function generateKeyPairFromSeed(seed: Uint8Array): Promise<KeyPair> {
  if (seed.length < 32) {
    throw new Error('Seed must be at least 32 bytes');
  }
  
  // Derive private key from seed
  const privateKey = await deriveKey(seed, new Uint8Array(16), 10000, PRIVATE_KEY_SIZE);
  
  // Derive public key
  const publicKey = await derivePublicKey(privateKey);
  
  return {
    privateKey,
    publicKey
  };
}

/**
 * Generates multiple key pairs in batch
 * @param count - Number of key pairs to generate
 * @returns Promise resolving to array of key pairs
 */
export async function batchGenerateKeyPairs(count: number): Promise<KeyPair[]> {
  if (count <= 0) {
    throw new Error('Count must be positive');
  }
  
  const promises: Promise<KeyPair>[] = [];
  for (let i = 0; i < count; i++) {
    promises.push(generateKeyPair());
  }
  
  return Promise.all(promises);
}

/**
 * Validates a key pair by checking if public key matches private key
 * @param keyPair - Key pair to validate
 * @returns Promise resolving to true if valid
 */
export async function validateKeyPair(keyPair: KeyPair): Promise<boolean> {
  try {
    validateSize(keyPair.privateKey, PRIVATE_KEY_SIZE, 'private key');
    validateSize(keyPair.publicKey, PUBLIC_KEY_SIZE, 'public key');
    
    const derivedPublic = await derivePublicKey(keyPair.privateKey);
    return constantTimeEqual(keyPair.publicKey, derivedPublic);
  } catch {
    return false;
  }
}

/**
 * Derives a child key pair from a parent private key and index
 * @param parentPrivateKey - Parent private key
 * @param index - Child key index
 * @returns Promise resolving to child key pair
 */
export async function deriveChildKeyPair(parentPrivateKey: PrivateKey, index: number): Promise<KeyPair> {
  validateSize(parentPrivateKey, PRIVATE_KEY_SIZE, 'parent private key');
  
  // Create derivation data
  const indexBytes = new Uint8Array(4);
  new DataView(indexBytes.buffer).setUint32(0, index, false);
  
  // Combine parent key and index
  const derivationData = new Uint8Array(parentPrivateKey.length + indexBytes.length);
  derivationData.set(parentPrivateKey);
  derivationData.set(indexBytes, parentPrivateKey.length);
  
  // Derive child private key
  const childPrivateKey = await computeHash(derivationData);
  validateSize(childPrivateKey, PRIVATE_KEY_SIZE, 'child private key');
  
  // Derive child public key
  const childPublicKey = await derivePublicKey(childPrivateKey);
  
  return {
    privateKey: childPrivateKey,
    publicKey: childPublicKey
  };
}

/**
 * Generates an HD (Hierarchical Deterministic) wallet
 * @param masterSeed - Master seed for the wallet
 * @param count - Number of key pairs to generate
 * @returns Promise resolving to array of key pairs
 */
export async function generateHDWallet(masterSeed: Uint8Array, count: number): Promise<KeyPair[]> {
  if (count <= 0) {
    throw new Error('Count must be positive');
  }
  
  // Generate master key pair
  const masterKeyPair = await generateKeyPairFromSeed(masterSeed);
  
  // Generate child key pairs
  const keyPairs: KeyPair[] = [masterKeyPair];
  
  for (let i = 1; i < count; i++) {
    const childKeyPair = await deriveChildKeyPair(masterKeyPair.privateKey, i);
    keyPairs.push(childKeyPair);
  }
  
  return keyPairs;
}

/**
 * Derives a key pair from a password
 * @param password - Password string
 * @param salt - Salt for key derivation
 * @param iterations - PBKDF2 iterations (default: 100000)
 * @returns Promise resolving to key pair
 */
export async function deriveKeyPairFromPassword(
  password: string,
  salt: Uint8Array,
  iterations: number = 100000
): Promise<KeyPair> {
  const passwordBytes = new TextEncoder().encode(password);
  const privateKey = await deriveKey(passwordBytes, salt, iterations, PRIVATE_KEY_SIZE);
  const publicKey = await derivePublicKey(privateKey);
  
  return {
    privateKey,
    publicKey
  };
}

/**
 * Securely erases a key pair from memory
 * @param keyPair - Key pair to erase
 */
export function secureEraseKeyPair(keyPair: KeyPair): void {
  secureZero(keyPair.privateKey);
  secureZero(keyPair.publicKey);
}

/**
 * Creates a backup of a key pair (copies the data)
 * @param keyPair - Key pair to backup
 * @returns New key pair with copied data
 */
export function backupKeyPair(keyPair: KeyPair): KeyPair {
  return {
    privateKey: copyBytes(keyPair.privateKey),
    publicKey: copyBytes(keyPair.publicKey)
  };
}

/**
 * Serializes a key pair to a portable format
 * @param keyPair - Key pair to serialize
 * @returns Serialized key pair as JSON string
 */
export function serializeKeyPair(keyPair: KeyPair): string {
  return JSON.stringify({
    privateKey: Array.from(keyPair.privateKey),
    publicKey: Array.from(keyPair.publicKey)
  });
}

/**
 * Deserializes a key pair from a portable format
 * @param serialized - Serialized key pair JSON string
 * @returns Deserialized key pair
 * @throws Error if deserialization fails
 */
export function deserializeKeyPair(serialized: string): KeyPair {
  try {
    const data = JSON.parse(serialized);
    
    if (!data.privateKey || !data.publicKey) {
      throw new Error('Missing key data');
    }
    
    const privateKey = new Uint8Array(data.privateKey);
    const publicKey = new Uint8Array(data.publicKey);
    
    validateSize(privateKey, PRIVATE_KEY_SIZE, 'private key');
    validateSize(publicKey, PUBLIC_KEY_SIZE, 'public key');
    
    return { privateKey, publicKey };
  } catch (error) {
    throw new Error(`Failed to deserialize key pair: ${error}`);
  }
}

File: go/keypair.go
package topayz512

import (
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"sync"
	"time"
)

// Key pair generation and management for TOPAY-Z512 with optimizations

// KeyPair represents a public/private key pair
type KeyPair struct {
	PublicKey  *PublicKey  `json:"public_key"`
	PrivateKey *PrivateKey `json:"private_key"`
	Timestamp  time.Time   `json:"timestamp"`
	KeyID      []byte      `json:"key_id"`
}

// GenerateKeyPair generates a new cryptographic key pair
func GenerateKeyPair() (PrivateKey, PublicKey, error) {
	// Generate random private key
	privateBytes, err := SecureRandom(PrivateKeySize)
	if err != nil {
		return PrivateKey{}, PublicKey{}, err
	}

	var privateKey PrivateKey
	copy(privateKey[:], privateBytes)

	// Derive public key from private key
	publicKey := DerivePublicKey(privateKey)

	return privateKey, publicKey, nil
}

// GenerateKeyPairAdvanced generates a new TOPAY-Z512 key pair with optimizations
func GenerateKeyPairAdvanced() (*KeyPair, error) {
	// Use pooled buffers for key generation
	privateKeyData := GetBuffer(PrivateKeySize)
	publicKeyData := GetBuffer(PublicKeySize)
	keyID := GetBuffer(16) // 128-bit key ID

	// Generate secure random private key
	if _, err := rand.Read(privateKeyData); err != nil {
		PutBuffer(privateKeyData)
		PutBuffer(publicKeyData)
		PutBuffer(keyID)
		return nil, err
	}

	// Derive public key from private key using optimized operations
	if err := derivePublicKeyAdvanced(publicKeyData, privateKeyData); err != nil {
		// Secure erase private key on error
		SecureZero(privateKeyData)
		PutBuffer(privateKeyData)
		PutBuffer(publicKeyData)
		PutBuffer(keyID)
		return nil, err
	}

	// Generate unique key ID
	if _, err := rand.Read(keyID); err != nil {
		SecureZero(privateKeyData)
		PutBuffer(privateKeyData)
		PutBuffer(publicKeyData)
		PutBuffer(keyID)
		return nil, err
	}

	timestamp := time.Now()

	// Create public key
	var pubKey PublicKey
	copy(pubKey[:], publicKeyData)

	// Create private key
	var privKey PrivateKey
	copy(privKey[:], privateKeyData)

	// Return buffers to pool
	PutBuffer(privateKeyData)
	PutBuffer(publicKeyData)
	PutBuffer(keyID)

	return &KeyPair{
		PublicKey:  &pubKey,
		PrivateKey: &privKey,
		Timestamp:  timestamp,
		KeyID:      make([]byte, 16), // Copy keyID data
	}, nil
}

// derivePublicKeyAdvanced derives the public key from private key using optimized operations
func derivePublicKeyAdvanced(publicKey, privateKey []byte) error {
	if len(publicKey) != PublicKeySize || len(privateKey) != PrivateKeySize {
		return errors.New("invalid key sizes")
	}

	// Use optimized hash-based key derivation
	hash := ComputeHash(privateKey)

	// Use SIMD operations for key derivation when available
	if simdCaps.SSE2 && len(publicKey) >= 16 {
		// Process in 16-byte chunks using vectorized operations
		for i := 0; i < len(publicKey); i += 16 {
			end := i + 16
			if end > len(publicKey) {
				end = len(publicKey)
			}

			// Use hash bytes cyclically for key derivation
			hashOffset := (i / 16) % (HashSize / 16)
			hashChunk := hash[hashOffset*16 : hashOffset*16+16]

			if end-i == 16 {
				VectorizedXOR(publicKey[i:end], privateKey[i:end], hashChunk)
			} else {
				// Fallback for partial chunks
				for j := i; j < end; j++ {
					publicKey[j] = privateKey[j] ^ hashChunk[j-i]
				}
			}
		}
	} else {
		// Fallback to scalar operations
		for i := 0; i < len(publicKey); i++ {
			hashByte := hash[i%HashSize]
			publicKey[i] = privateKey[i] ^ hashByte
		}
	}

	return nil
}

// DerivePublicKey derives a public key from a private key
func DerivePublicKey(privateKey PrivateKey) PublicKey {
	// Use SHA-256 based key derivation
	// In production, use proper elliptic curve or lattice-based derivation
	hasher := sha256.New()
	hasher.Write(privateKey[:])
	hasher.Write([]byte("TOPAY-Z512-PUBLIC-KEY-DERIVATION"))

	firstHash := hasher.Sum(nil)

	// Second round for additional security
	hasher.Reset()
	hasher.Write(firstHash)
	hasher.Write(privateKey[:])
	secondHash := hasher.Sum(nil)

	var publicKey PublicKey
	copy(publicKey[:], secondHash)

	return publicKey
}

// GenerateKeyPairFromSeed generates a deterministic key pair from a seed
func GenerateKeyPairFromSeed(seed []byte) (PrivateKey, PublicKey, error) {
	if len(seed) < 32 {
		return PrivateKey{}, PublicKey{}, ErrInvalidKeySize
	}

	// Derive private key from seed
	hasher := sha256.New()
	hasher.Write(seed)
	hasher.Write([]byte("TOPAY-Z512-PRIVATE-KEY-SEED"))

	privateHash := hasher.Sum(nil)

	// Extend to full private key size
	var privateKey PrivateKey
	for i := 0; i < PrivateKeySize; i++ {
		privateKey[i] = privateHash[i%len(privateHash)]
	}

	// Derive public key
	publicKey := DerivePublicKey(privateKey)

	return privateKey, publicKey, nil
}

// VerifyKeyPair verifies that a private and public key form a valid pair
func VerifyKeyPair(privateKey PrivateKey, publicKey PublicKey) bool {
	derivedPublic := DerivePublicKey(privateKey)
	return ConstantTimeEqual(publicKey[:], derivedPublic[:])
}

// IsValidPrivateKey checks if a private key is valid
func IsValidPrivateKey(privateKey PrivateKey) bool {
	// Check if private key is not all zeros
	var zero PrivateKey
	if ConstantTimeEqual(privateKey[:], zero[:]) {
		return false
	}

	// Check if private key is not all ones
	var ones PrivateKey
	for i := range ones {
		ones[i] = 0xFF
	}
	return !ConstantTimeEqual(privateKey[:], ones[:])
}

// IsValidPublicKey checks if a public key is valid
func IsValidPublicKey(publicKey PublicKey) bool {
	// Check if public key is not all zeros
	var zero PublicKey
	if ConstantTimeEqual(publicKey[:], zero[:]) {
		return false
	}

	// Check if public key is not all ones
	var ones PublicKey
	for i := range ones {
		ones[i] = 0xFF
	}
	return !ConstantTimeEqual(publicKey[:], ones[:])
}

// Batch key pair generation

// BatchKeyPairResult represents the result of batch key pair generation
type BatchKeyPairResult struct {
	Index      int
	PrivateKey PrivateKey
	PublicKey  PublicKey
	Error      error
}

// BatchGenerateKeyPairs generates multiple key pairs in parallel
func BatchGenerateKeyPairs(count int) ([]PrivateKey, []PublicKey, error) {
	if count <= 0 {
		return nil, nil, ErrInvalidFragmentCount
	}

	privateKeys := make([]PrivateKey, count)
	publicKeys := make([]PublicKey, count)

	// Use optimal number of goroutines
	numWorkers := OptimalThreadCount()
	if numWorkers > count {
		numWorkers = count
	}

	// Channel for work distribution
	workChan := make(chan int, count)
	resultChan := make(chan BatchKeyPairResult, count)

	// Start workers
	var wg sync.WaitGroup
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for index := range workChan {
				privateKey, publicKey, err := GenerateKeyPair()
				resultChan <- BatchKeyPairResult{
					Index:      index,
					PrivateKey: privateKey,
					PublicKey:  publicKey,
					Error:      err,
				}
			}
		}()
	}

	// Send work
	go func() {
		for i := 0; i < count; i++ {
			workChan <- i
		}
		close(workChan)
	}()

	// Wait for workers to complete
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Collect results
	for result := range resultChan {
		if result.Error != nil {
			return nil, nil, result.Error
		}
		privateKeys[result.Index] = result.PrivateKey
		publicKeys[result.Index] = result.PublicKey
	}

	return privateKeys, publicKeys, nil
}

// BatchGenerateKeyPairsFromSeeds generates key pairs from multiple seeds in parallel
func BatchGenerateKeyPairsFromSeeds(seeds [][]byte) ([]PrivateKey, []PublicKey, error) {
	if len(seeds) == 0 {
		return nil, nil, ErrEmptyData
	}

	privateKeys := make([]PrivateKey, len(seeds))
	publicKeys := make([]PublicKey, len(seeds))

	// Use optimal number of goroutines
	numWorkers := OptimalThreadCount()
	if numWorkers > len(seeds) {
		numWorkers = len(seeds)
	}

	// Channel for work distribution
	workChan := make(chan int, len(seeds))
	resultChan := make(chan BatchKeyPairResult, len(seeds))

	// Start workers
	var wg sync.WaitGroup
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for index := range workChan {
				privateKey, publicKey, err := GenerateKeyPairFromSeed(seeds[index])
				resultChan <- BatchKeyPairResult{
					Index:      index,
					PrivateKey: privateKey,
					PublicKey:  publicKey,
					Error:      err,
				}
			}
		}()
	}

	// Send work
	go func() {
		for i := range seeds {
			workChan <- i
		}
		close(workChan)
	}()

	// Wait for workers to complete
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Collect results
	for result := range resultChan {
		if result.Error != nil {
			return nil, nil, result.Error
		}
		privateKeys[result.Index] = result.PrivateKey
		publicKeys[result.Index] = result.PublicKey
	}

	return privateKeys, publicKeys, nil
}

// Key pair utilities

// KeyPairEqual compares two key pairs for equality
func KeyPairEqual(kp1, kp2 KeyPair) bool {
	return ConstantTimeEqual((*kp1.PrivateKey)[:], (*kp2.PrivateKey)[:]) &&
		ConstantTimeEqual((*kp1.PublicKey)[:], (*kp2.PublicKey)[:])
}

// PrivateKeyEqual compares two private keys for equality
func PrivateKeyEqual(pk1, pk2 PrivateKey) bool {
	return ConstantTimeEqual(pk1[:], pk2[:])
}

// PublicKeyEqual compares two public keys for equality
func PublicKeyEqual(pk1, pk2 PublicKey) bool {
	return ConstantTimeEqual(pk1[:], pk2[:])
}

// SecureErasePrivateKey securely erases a private key from memory
func SecureErasePrivateKey(privateKey *PrivateKey) {
	SecureZero(privateKey[:])
}

// SecureEraseKeyPair securely erases a key pair from memory
func SecureEraseKeyPair(keyPair *KeyPair) {
	SecureZero((*keyPair.PrivateKey)[:])
	SecureZero((*keyPair.PublicKey)[:])
}

// Key derivation functions

// DeriveKeyFromPassword derives a private key from a password using PBKDF2
func DeriveKeyFromPassword(password, salt []byte, iterations int) (PrivateKey, error) {
	if len(password) == 0 {
		return PrivateKey{}, ErrEmptyData
	}

	if len(salt) < 16 {
		return PrivateKey{}, ErrInvalidKeySize
	}

	// Simple PBKDF2-like derivation
	derived := make([]byte, PrivateKeySize)

	current := append(password, salt...)
	for i := 0; i < iterations; i++ {
		hash := ComputeHash(current)
		current = hash[:]
	}

	copy(derived, current[:PrivateKeySize])

	var privateKey PrivateKey
	copy(privateKey[:], derived)

	// Ensure the derived key is valid
	if !IsValidPrivateKey(privateKey) {
		// If invalid, hash again with a different salt
		modifiedSalt := append(salt, 0x01)
		return DeriveKeyFromPassword(password, modifiedSalt, iterations)
	}

	return privateKey, nil
}

// DeriveChildKey derives a child key from a parent private key and index
func DeriveChildKey(parentKey PrivateKey, index uint32) PrivateKey {
	// Simple child key derivation
	hasher := sha256.New()
	hasher.Write(parentKey[:])
	hasher.Write([]byte("TOPAY-Z512-CHILD-KEY"))

	// Add index as bytes
	indexBytes := make([]byte, 4)
	indexBytes[0] = byte(index >> 24)
	indexBytes[1] = byte(index >> 16)
	indexBytes[2] = byte(index >> 8)
	indexBytes[3] = byte(index)
	hasher.Write(indexBytes)

	childHash := hasher.Sum(nil)

	var childKey PrivateKey
	copy(childKey[:], childHash)

	return childKey
}

// Key pair performance benchmarking

// KeyPairBenchmark represents key pair generation performance metrics
type KeyPairBenchmark struct {
	KeyPairsPerSec    float64
	AvgLatencyMs      float64
	BatchSpeedupRatio float64
}

// BenchmarkKeyPairGeneration measures key pair generation performance
func BenchmarkKeyPairGeneration(iterations int) KeyPairBenchmark {
	// Single key pair generation benchmark
	start := time.Now()
	for i := 0; i < iterations; i++ {
		_, _, _ = GenerateKeyPair()
	}
	singleDuration := time.Since(start)

	// Batch key pair generation benchmark
	start = time.Now()
	_, _, _ = BatchGenerateKeyPairs(iterations)
	batchDuration := time.Since(start)

	keyPairsPerSec := float64(iterations) / singleDuration.Seconds()
	avgLatencyMs := singleDuration.Seconds() * 1000 / float64(iterations)
	batchSpeedupRatio := singleDuration.Seconds() / batchDuration.Seconds()

	return KeyPairBenchmark{
		KeyPairsPerSec:    keyPairsPerSec,
		AvgLatencyMs:      avgLatencyMs,
		BatchSpeedupRatio: batchSpeedupRatio,
	}
}

// Advanced key pair operations

// RecoverPublicKey attempts to recover a public key from a private key
func RecoverPublicKey(privateKey PrivateKey) (PublicKey, error) {
	if !IsValidPrivateKey(privateKey) {
		return PublicKey{}, ErrInvalidKeySize
	}

	publicKey := DerivePublicKey(privateKey)
	return publicKey, nil
}

// ValidateKeyPairIntegrity performs comprehensive validation of a key pair
func ValidateKeyPairIntegrity(keyPair KeyPair) error {
	// Check private key validity
	if !IsValidPrivateKey(*keyPair.PrivateKey) {
		return ErrInvalidKeySize
	}

	// Check public key validity
	if !IsValidPublicKey(*keyPair.PublicKey) {
		return ErrInvalidKeySize
	}

	// Verify key pair consistency
	if !VerifyKeyPair(*keyPair.PrivateKey, *keyPair.PublicKey) {
		return ErrInvalidKeySize
	}

	return nil
}

// GenerateHDWallet generates a hierarchical deterministic wallet
func GenerateHDWallet(seed []byte, depth int) ([]KeyPair, error) {
	if depth <= 0 || depth > 256 {
		return nil, ErrInvalidFragmentCount
	}

	masterPrivate, masterPublic, err := GenerateKeyPairFromSeed(seed)
	if err != nil {
		return nil, err
	}

	keyPairs := make([]KeyPair, depth)
	keyPairs[0] = KeyPair{PrivateKey: &masterPrivate, PublicKey: &masterPublic}

	currentPrivate := masterPrivate
	for i := 1; i < depth; i++ {
		childPrivate := DeriveChildKey(currentPrivate, uint32(i))
		childPublic := DerivePublicKey(childPrivate)

		keyPairs[i] = KeyPair{PrivateKey: &childPrivate, PublicKey: &childPublic}
		currentPrivate = childPrivate
	}

	return keyPairs, nil
}
